tsdb/importer/importer.go

```diff
@@ -0,0 +1,583 @@
+
+// ImportFromFile imports data from a file formatted according to the Open Metrics format,
+// converts it into block(s), and places the newly created block(s) in the
+// TSDB DB directory, where it is treated like any other block.
+func ImportFromFile(r io.Reader, dbPath string, maxSamplesInMemory int, logger log.Logger) error {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	level.Debug(logger).Log("msg", "creating temp directory")
+	tmpDbDir, err := ioutil.TempDir("", "importer")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDbDir)
+
+	level.Debug(logger).Log("msg", "reading existing block metadata to infer time ranges")
+	dbMint, dbMaxt, blockTimes, err := getDbTimes(dbPath)
+	if err != nil {
+		return err
+	}
+
+	level.Debug(logger).Log("msg", "creating new block skeletons based on existing blocks")
+	blockMetas := constructBlockMetadata(dbMint, dbMaxt, blockTimes)
+
+	level.Info(logger).Log("msg", "importing input data and aligning with existing DB")
+	if err = writeSamples(r, tmpDbDir, dbMint, dbMaxt, blockMetas, maxSamplesInMemory, logger); err != nil {
+		return err
+	}
+
+	level.Info(logger).Log("msg", "merging fully overlapping blocks")
+	if err = mergeBlocks(tmpDbDir, blockMetas, logger); err != nil {
+		return err
+	}
+
+	level.Debug(logger).Log("msg", "copying newly created blocks from temp location to actual DB location")
+	if err = fileutil.CopyDirs(tmpDbDir, dbPath); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// writeSamples parses each metric sample, and writes the samples to the correspondingly aligned block.
+// It uses the block indexes assigned to each sample, and block meta info, gathered in collectSampleInformation().
+func writeSamples(r io.Reader, dbDir string, dbMint, dbMaxt int64, metas []*newBlockMeta, maxSamplesInMemory int, logger log.Logger) error {
+	blockTimes := make([]blockTimestampPair, 0)
+	for _, meta := range metas {
+		blockTimes = append(blockTimes, blockTimestampPair{
+			start: meta.mint,
+			end:   meta.maxt,
+		})
+	}
+
+	blocks := getEmptyBlocks(len(metas))
+	currentPassCount := 0
+
+	// TODO: Maybe use bufio.Reader.Readline instead?
+	// https://stackoverflow.com/questions/21124327/how-to-read-a-text-file-line-by-line-in-go-when-some-lines-are-long-enough-to-ca
+	// Use a streaming approach to avoid loading too much data at once.
+	scanner := bufio.NewScanner(r)
+	buf := new(bytes.Buffer)
+	scanner.Split(sampleStreamer(buf))
+
+	for scanner.Scan() {
+		if currentPassCount == 0 {
+			blocks = getEmptyBlocks(len(metas))
+		}
+
+		encSample := scanner.Bytes()
+		decBuf := bytes.NewBuffer(encSample)
+		sample := tsdb.MetricSample{}
+		if err := gob.NewDecoder(decBuf).Decode(&sample); err != nil {
+			level.Error(logger).Log("msg", "failed to decode current entry returned by file scanner", "err", err)
+			return err
+		}
+
+		// Find what block does this sample belong to.
+		var blockIndex int
+		if len(metas) == 1 || sample.Timestamp < dbMint {
```

Comment:
you can just check metas[0] and metas[len(metas)-1] it would be more obvious.

Basically any sample outside the mint goes to the last time range and outside maxt goes in the first time range